<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Libraries" data-toc-modified-id="Libraries-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Libraries</a></span></li><li><span><a href="#Dataset" data-toc-modified-id="Dataset-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Dataset</a></span></li><li><span><a href="#Exploring-the-data-set" data-toc-modified-id="Exploring-the-data-set-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Exploring the data-set</a></span></li><li><span><a href="#Missing-Values" data-toc-modified-id="Missing-Values-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Missing Values</a></span><ul class="toc-item"><li><span><a href="#Deletion" data-toc-modified-id="Deletion-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Deletion</a></span></li><li><span><a href="#Deduction" data-toc-modified-id="Deduction-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Deduction</a></span><ul class="toc-item"><li><span><a href="#Major-Discipline" data-toc-modified-id="Major-Discipline-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Major Discipline</a></span></li><li><span><a href="#Experience" data-toc-modified-id="Experience-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Experience</a></span></li><li><span><a href="#Company-Size-and-Company-Type" data-toc-modified-id="Company-Size-and-Company-Type-4.2.3"><span class="toc-item-num">4.2.3&nbsp;&nbsp;</span>Company Size and Company Type</a></span></li><li><span><a href="#Last-New-Job" data-toc-modified-id="Last-New-Job-4.2.4"><span class="toc-item-num">4.2.4&nbsp;&nbsp;</span>Last New Job</a></span></li></ul></li><li><span><a href="#Prediction" data-toc-modified-id="Prediction-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Prediction</a></span><ul class="toc-item"><li><span><a href="#Gender" data-toc-modified-id="Gender-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Gender</a></span></li><li><span><a href="#Enrolled-University" data-toc-modified-id="Enrolled-University-4.3.2"><span class="toc-item-num">4.3.2&nbsp;&nbsp;</span>Enrolled University</a></span></li><li><span><a href="#Education-Level" data-toc-modified-id="Education-Level-4.3.3"><span class="toc-item-num">4.3.3&nbsp;&nbsp;</span>Education Level</a></span></li><li><span><a href="#Major-Discipline" data-toc-modified-id="Major-Discipline-4.3.4"><span class="toc-item-num">4.3.4&nbsp;&nbsp;</span>Major Discipline</a></span></li><li><span><a href="#Experience" data-toc-modified-id="Experience-4.3.5"><span class="toc-item-num">4.3.5&nbsp;&nbsp;</span>Experience</a></span></li><li><span><a href="#Company-Size" data-toc-modified-id="Company-Size-4.3.6"><span class="toc-item-num">4.3.6&nbsp;&nbsp;</span>Company Size</a></span></li><li><span><a href="#Company-Type" data-toc-modified-id="Company-Type-4.3.7"><span class="toc-item-num">4.3.7&nbsp;&nbsp;</span>Company Type</a></span></li><li><span><a href="#Last-New-Job" data-toc-modified-id="Last-New-Job-4.3.8"><span class="toc-item-num">4.3.8&nbsp;&nbsp;</span>Last New Job</a></span></li></ul></li></ul></li><li><span><a href="#Re-building" data-toc-modified-id="Re-building-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Re-building</a></span></li></ul></div>

## Libraries

In [1]:
import pandas as pd
import numpy as np
import datawig

#warnings
import warnings
warnings.filterwarnings('ignore')

## Dataset
The source of the data is kaggle website: [Link](https://www.kaggle.com/arashnic/hr-analytics-job-change-of-data-scientists).

In [2]:
file_path = 'data/aug_train.csv'
dataFrame = pd.read_csv(file_path)
dataFrame.head(5)

enrollee_id      city  city_development_index gender  \
0         8949  city_103                   0.920   Male   
1        29725   city_40                   0.776   Male   
2        11561   city_21                   0.624    NaN   
3        33241  city_115                   0.789    NaN   
4          666  city_162                   0.767   Male   

       relevent_experience enrolled_university education_level  \
0  Has relevent experience       no_enrollment        Graduate   
1   No relevent experience       no_enrollment        Graduate   
2   No relevent experience    Full time course        Graduate   
3   No relevent experience                 NaN        Graduate   
4  Has relevent experience       no_enrollment         Masters   

  major_discipline experience company_size    company_type last_new_job  \
0             STEM        >20          NaN             NaN            1   
1             STEM         15        50-99         Pvt Ltd           >4   
2             STEM          5          NaN             NaN        never   
3  Business Degree         <1          NaN         Pvt Ltd        never   
4             STEM        >20        50-99  Funded Startup            4   

   training_hours  target  
0              36     1.0  
1              47     0.0  
2              83     0.0  
3              52     1.0  
4               8     0.0

In [3]:
# replacing values of relevant experience column with 
# 1 (= Has relevent experience) and 0 (= No relevent experience)

dataFrame['relevent_experience'] = dataFrame['relevent_experience'].replace(['Has relevent experience'],1)
dataFrame['relevent_experience'] = dataFrame['relevent_experience'].replace(['No relevent experience'],0)

In [4]:
dataFrame.head(5)

enrollee_id      city  city_development_index gender  relevent_experience  \
0         8949  city_103                   0.920   Male                    1   
1        29725   city_40                   0.776   Male                    0   
2        11561   city_21                   0.624    NaN                    0   
3        33241  city_115                   0.789    NaN                    0   
4          666  city_162                   0.767   Male                    1   

  enrolled_university education_level major_discipline experience  \
0       no_enrollment        Graduate             STEM        >20   
1       no_enrollment        Graduate             STEM         15   
2    Full time course        Graduate             STEM          5   
3                 NaN        Graduate  Business Degree         <1   
4       no_enrollment         Masters             STEM        >20   

  company_size    company_type last_new_job  training_hours  target  
0          NaN             NaN            1              36     1.0  
1        50-99         Pvt Ltd           >4              47     0.0  
2          NaN             NaN        never              83     0.0  
3          NaN         Pvt Ltd        never              52     1.0  
4        50-99  Funded Startup            4               8     0.0

This dataset is named <b>dataFrame</b> and includes 14 columns and 19158 rows.

## Exploring the data-set
Since our prediction model is as accurate as our data, the dataset needs to be studied carefully. There may be features that should be added, removed, or aggregated, the type of some of them may need to change (i.e. categorical to numerical) or there could be some missing data that must be dealt with.
The first step is studying the features:

In [5]:
dataFrame.columns

Index(['enrollee_id', 'city', 'city_development_index', 'gender',
       'relevent_experience', 'enrolled_university', 'education_level',
       'major_discipline', 'experience', 'company_size', 'company_type',
       'last_new_job', 'training_hours', 'target'],
      dtype='object')

Looking at the elements of each column is a good way to understand the data better and determine any incorrect values.

In [6]:
#make sure every subject has a unique ID 

dataFrame['enrollee_id'].value_counts()

2047     1
2692     1
4743     1
27272    1
25225    1
        ..
17682    1
19731    1
29972    1
25878    1
6147     1
Name: enrollee_id, Length: 19158, dtype: int64

In [7]:
#count values of city 

dataFrame['city'].value_counts()

city_103    4355
city_21     2702
city_16     1533
city_114    1336
city_160     845
            ... 
city_111       3
city_121       3
city_129       3
city_140       1
city_171       1
Name: city, Length: 123, dtype: int64

In [8]:
#count values of city development index 

dataFrame['city_development_index'].value_counts()

0.920    5200
0.624    2702
0.910    1533
0.926    1336
0.698     683
         ... 
0.649       4
0.807       4
0.781       3
0.625       3
0.664       1
Name: city_development_index, Length: 93, dtype: int64

Since there are a great number of values in this column, checking each of them is difficult. Looking at the type of this column helps to know if there is a string in the column that is supposed to include only numbers:

In [9]:
# explore city

dataFrame['city']

0        city_103
1         city_40
2         city_21
3        city_115
4        city_162
           ...   
19153    city_173
19154    city_103
19155    city_103
19156     city_65
19157     city_67
Name: city, Length: 19158, dtype: object

In [10]:
#explore city development index

dataFrame['city_development_index']

0        0.920
1        0.776
2        0.624
3        0.789
4        0.767
         ...  
19153    0.878
19154    0.920
19155    0.920
19156    0.802
19157    0.855
Name: city_development_index, Length: 19158, dtype: float64

In [11]:
#count values of gender

dataFrame['gender'].value_counts()

Male      13221
Female     1238
Other       191
Name: gender, dtype: int64

In [12]:
#count values of relevent experience

dataFrame['relevent_experience'].value_counts()

1    13792
0     5366
Name: relevent_experience, dtype: int64

In [13]:
#count values of enrolled university

dataFrame['enrolled_university'].value_counts()

no_enrollment       13817
Full time course     3757
Part time course     1198
Name: enrolled_university, dtype: int64

In [14]:
#count values of education level

dataFrame['education_level'].value_counts()

Graduate          11598
Masters            4361
High School        2017
Phd                 414
Primary School      308
Name: education_level, dtype: int64

In [15]:
#count values of major discipline

dataFrame['major_discipline'].value_counts()

STEM               14492
Humanities           669
Other                381
Business Degree      327
Arts                 253
No Major             223
Name: major_discipline, dtype: int64

In [16]:
#count values of experience

dataFrame['experience'].value_counts()

>20    3286
5      1430
4      1403
3      1354
6      1216
2      1127
7      1028
10      985
9       980
8       802
15      686
11      664
14      586
1       549
<1      522
16      508
12      494
13      399
17      342
19      304
18      280
20      148
Name: experience, dtype: int64

In [17]:
#count values of company size

dataFrame['company_size'].value_counts()

50-99        3083
100-500      2571
10000+       2019
10/49        1471
1000-4999    1328
<10          1308
500-999       877
5000-9999     563
Name: company_size, dtype: int64

In [18]:
#count values of company type

dataFrame['company_type'].value_counts()

Pvt Ltd                9817
Funded Startup         1001
Public Sector           955
Early Stage Startup     603
NGO                     521
Other                   121
Name: company_type, dtype: int64

In [19]:
#count values of last new job

dataFrame['last_new_job'].value_counts()

1        8040
>4       3290
2        2900
never    2452
4        1029
3        1024
Name: last_new_job, dtype: int64

In [20]:
#count values of training hours

dataFrame['training_hours'].value_counts()

28     329
12     292
18     291
22     282
50     279
      ... 
266      6
272      5
234      5
286      5
238      4
Name: training_hours, Length: 241, dtype: int64

In [21]:
#explore training hours

dataFrame['training_hours']

0         36
1         47
2         83
3         52
4          8
        ... 
19153     42
19154     52
19155     44
19156     97
19157    127
Name: training_hours, Length: 19158, dtype: int64

In [22]:
#count values of target

dataFrame['target'].value_counts()

0.0    14381
1.0     4777
Name: target, dtype: int64

In [23]:
#type of values

dataFrame_info = dataFrame.info();

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19158 entries, 0 to 19157
Data columns (total 14 columns):
enrollee_id               19158 non-null int64
city                      19158 non-null object
city_development_index    19158 non-null float64
gender                    14650 non-null object
relevent_experience       19158 non-null int64
enrolled_university       18772 non-null object
education_level           18698 non-null object
major_discipline          16345 non-null object
experience                19093 non-null object
company_size              13220 non-null object
company_type              13018 non-null object
last_new_job              18735 non-null object
training_hours            19158 non-null int64
target                    19158 non-null float64
dtypes: float64(2), int64(3), object(9)
memory usage: 2.0+ MB


In [24]:
dataFrame.describe()

enrollee_id  city_development_index  relevent_experience  \
count  19158.000000            19158.000000         19158.000000   
mean   16875.358179                0.828848             0.719908   
std     9616.292592                0.123362             0.449056   
min        1.000000                0.448000             0.000000   
25%     8554.250000                0.740000             0.000000   
50%    16982.500000                0.903000             1.000000   
75%    25169.750000                0.920000             1.000000   
max    33380.000000                0.949000             1.000000   

       training_hours        target  
count    19158.000000  19158.000000  
mean        65.366896      0.249348  
std         60.058462      0.432647  
min          1.000000      0.000000  
25%         23.000000      0.000000  
50%         47.000000      0.000000  
75%         88.000000      0.000000  
max        336.000000      1.000000

## Missing Values

In [25]:
missing_data = pd.DataFrame(dataFrame.isna().sum(), columns=[
                            'Number of missing values'])
missing_data['Percentage of missing values'] = missing_data['Number of missing values'].aggregate(
    lambda x: (x/dataFrame.shape[0])*100)
missing_data

Number of missing values  Percentage of missing values
enrollee_id                                    0                      0.000000
city                                           0                      0.000000
city_development_index                         0                      0.000000
gender                                      4508                     23.530640
relevent_experience                            0                      0.000000
enrolled_university                          386                      2.014824
education_level                              460                      2.401086
major_discipline                            2813                     14.683161
experience                                    65                      0.339284
company_size                                5938                     30.994885
company_type                                6140                     32.049274
last_new_job                                 423                      2.207955
training_hours                                 0                      0.000000
target                                         0                      0.000000

In [26]:
len(dataFrame.dropna())

8955

Even if it would be more convenient to delete all the rows that contain missing values, we'd rather not doing it because half of the rows have at least one missing values. For this reason we decided to choose another strategy and try to fill them.

### Deletion

The first step was to delete all the rows that contained more than 6 missing values.

In [27]:
#save only the rows that have less than 6 missing values

dataFrame = dataFrame[dataFrame.isnull().sum(axis=1) < 6]

In [28]:
missing_data = pd.DataFrame(dataFrame.isna().sum(), columns=[
                            'Number of missing values'])
missing_data['Percentage of missing values'] = missing_data['Number of missing values'].aggregate(
    lambda x: (x/dataFrame.shape[0])*100)
missing_data

Number of missing values  Percentage of missing values
enrollee_id                                    0                      0.000000
city                                           0                      0.000000
city_development_index                         0                      0.000000
gender                                      4438                     23.255083
relevent_experience                            0                      0.000000
enrolled_university                          332                      1.739677
education_level                              392                      2.054077
major_discipline                            2739                     14.352337
experience                                    56                      0.293440
company_size                                5864                     30.727311
company_type                                6066                     31.785789
last_new_job                                 390                      2.043597
training_hours                                 0                      0.000000
target                                         0                      0.000000

### Deduction

Some of the features are related among them and missing values can be manually filled applying deduction.

#### Major Discipline

In [29]:
dataFrame[dataFrame['education_level'] == "Primary School"]["major_discipline"].isna().value_counts()

True    307
Name: major_discipline, dtype: int64

In [30]:
dataFrame[dataFrame['education_level'] == "High School"]["major_discipline"].isna().value_counts()

True    2013
Name: major_discipline, dtype: int64

The value of major discipline is always null if the education level is primary school or high school. This fact is reletad to the organizazion of educational system because neither primary school or high school have a major discipline. Thus, we decided to set all the values of major discipline corresponding to primary school and high school to *No Major*.

In [31]:
dataFrame['major_discipline'] = np.where(
    dataFrame['education_level'] == "Primary School", "No Major", dataFrame['major_discipline'])
dataFrame['major_discipline'] = np.where(
    dataFrame['education_level'] == "High School", "No Major", dataFrame['major_discipline'])

In [32]:
dataFrame[(dataFrame['major_discipline'].isna())]

enrollee_id      city  city_development_index gender  \
13            5826   city_21                   0.624   Male   
39           24659   city_71                   0.884   Male   
69            4830   city_90                   0.698    NaN   
136           3116   city_21                   0.624   Male   
153           8241   city_16                   0.910    NaN   
...            ...       ...                     ...    ...   
19095        15806  city_165                   0.903    NaN   
19100        16368   city_65                   0.802    NaN   
19118        12211  city_126                   0.479    NaN   
19142        13750   city_40                   0.776   Male   
19151        11385  city_149                   0.689   Male   

       relevent_experience enrolled_university education_level  \
13                       0                 NaN             NaN   
39                       0       no_enrollment             NaN   
69                       0                 NaN             NaN   
136                      0       no_enrollment             NaN   
153                      1       no_enrollment             NaN   
...                    ...                 ...             ...   
19095                    0       no_enrollment             NaN   
19100                    0       no_enrollment             NaN   
19118                    0       no_enrollment             NaN   
19142                    1       no_enrollment             NaN   
19151                    0    Full time course        Graduate   

      major_discipline experience company_size         company_type  \
13                 NaN          2          NaN                  NaN   
39                 NaN          3          NaN                  NaN   
69                 NaN          2          NaN              Pvt Ltd   
136                NaN          3          NaN                  NaN   
153                NaN         11          NaN                  NaN   
...                ...        ...          ...                  ...   
19095              NaN          3          NaN        Public Sector   
19100              NaN          4          NaN                  NaN   
19118              NaN          1          NaN                  NaN   
19142              NaN          5        10/49  Early Stage Startup   
19151              NaN          2          NaN                  NaN   

      last_new_job  training_hours  target  
13           never              24     0.0  
39           never             106     0.0  
69           never             228     1.0  
136              1              24     0.0  
153              1               4     0.0  
...            ...             ...     ...  
19095        never              62     0.0  
19100            2             304     0.0  
19118        never             316     1.0  
19142            1              26     0.0  
19151            1              60     0.0  

[419 rows x 14 columns]

We still have 419 rows in which the major discipline is missing. Since no more reasoning can be done, these values should be filled by using prediction.

#### Experience

If there is no information about 
- experience
- company size
- company type
- last new job

and the relevent experience is 0 we can assume that the subject has not experience at all. We can also assume that, if last new job is set to "never", then also in this case the subject should have no experience. 

In [33]:
dataFrame['experience'] = np.where(((dataFrame['experience'].isna()) &
                                    (dataFrame['company_size'].isna()) &
                                    (dataFrame['company_type'].isna()) &
                                    ((dataFrame['last_new_job'].isna()) | 
                                     (dataFrame['last_new_job'] == "never")) &
                                    (dataFrame['relevent_experience'] == 0)),
                                   "<1", dataFrame['experience'])

Again, since it is not possible to apply further reasoning, we have to use prediction algorithms.

#### Company Size and Company Type

Missing values of company size and type are, in some way, related. However, we could not use reasoning for them because there is no unique correspondance between them. 

#### Last New Job

In [34]:
dataFrame[dataFrame['last_new_job'].isna()]

enrollee_id      city  city_development_index gender  \
58           13478   city_21                   0.624    NaN   
64            9572   city_11                   0.550    NaN   
90           19128   city_61                   0.913   Male   
135          23947  city_103                   0.920    NaN   
180          27708   city_94                   0.698   Male   
...            ...       ...                     ...    ...   
18944         2678  city_104                   0.924    NaN   
19043        17966  city_114                   0.926    NaN   
19106        18131  city_100                   0.887   Male   
19113        11622   city_19                   0.682    NaN   
19139         5603   city_21                   0.624    NaN   

       relevent_experience enrolled_university education_level  \
58                       1    Full time course        Graduate   
64                       0    Full time course     High School   
90                       0    Full time course        Graduate   
135                      0       no_enrollment             Phd   
180                      0    Full time course        Graduate   
...                    ...                 ...             ...   
18944                    0       no_enrollment             Phd   
19043                    0    Full time course     High School   
19106                    0    Full time course         Masters   
19113                    0    Full time course        Graduate   
19139                    1    Full time course        Graduate   

      major_discipline experience company_size    company_type last_new_job  \
58                STEM          2        10/49  Funded Startup          NaN   
64            No Major          3          NaN             NaN          NaN   
90                STEM          4          NaN             NaN          NaN   
135               STEM         <1          NaN             NaN          NaN   
180               STEM          5    5000-9999             NaN          NaN   
...                ...        ...          ...             ...          ...   
18944             STEM        >20        50-99         Pvt Ltd          NaN   
19043         No Major          3          NaN             NaN          NaN   
19106             STEM          6      100-500   Public Sector          NaN   
19113             STEM          6          NaN             NaN          NaN   
19139             STEM          4          NaN             NaN          NaN   

       training_hours  target  
58                 32     1.0  
64                 98     0.0  
90                140     0.0  
135                70     0.0  
180                24     1.0  
...               ...     ...  
18944              27     0.0  
19043              20     0.0  
19106               8     0.0  
19113              53     0.0  
19139              13     0.0  

[390 rows x 14 columns]

The column "last_new_job" tells us how many years have last since that person has changed his job for the last time. That is, a person who is working for the first time has never changed his/her job. Thus, we replaced this type of missing values with *never*.

In [35]:
dataFrame['last_new_job'] = np.where((((dataFrame['experience'] == "<1") | 
                                       dataFrame['experience'].isna()) & 
                                      (dataFrame['last_new_job'].isna())), 
                                     "never", 
                                     dataFrame['last_new_job'])

### Prediction

For the prediction, we used all the columns except the one we were going to predict, the target column and the enrollee_id column. We did the prediction using DataWig, a framework for learning models to impute missing values in tables.

In [36]:
dataFrame.columns

Index(['enrollee_id', 'city', 'city_development_index', 'gender',
       'relevent_experience', 'enrolled_university', 'education_level',
       'major_discipline', 'experience', 'company_size', 'company_type',
       'last_new_job', 'training_hours', 'target'],
      dtype='object')

In [37]:
def prediction(df, input_columns, column_to_be_predicted, path='imputer_model', num_epochs=100):

    '''
    This function is used to predict missing values by using datawig. It takes as input
    a dataframe (with no missing values), the column(s) containing information about the 
    column we want to impute (list of name of those columns as strings), 
    the column we'd like to impute values for (name of the column as a string),
    the path for storing the imputer and the number of epochs to be done
    and returns an imputer that predicts a particular feature
    
    '''
    
    #Initialize a SimpleImputer model
    imputer = datawig.SimpleImputer(
    input_columns = input_columns, 
    output_column =  column_to_be_predicted, 
    output_path = path 
    )

    #Fit an imputer model on the train data
    imputer.fit(train_df=df, num_epochs=num_epochs)
    
    return imputer

#### Gender

In [38]:
df_not_missing_gender = dataFrame[dataFrame['gender'].isnull() == False]

df_train, df_test = datawig.utils.random_split(df_not_missing_gender, split_ratios=[0.8, 0.2])

input_columns_for_gender = ['city', 'city_development_index', 'relevent_experience',
       'enrolled_university', 'education_level', 'major_discipline',
       'experience', 'company_size', 'company_type', 'last_new_job',
       'training_hours']


gender_imputer_for_evaluation = prediction(df_train, 
                            input_columns_for_gender, 
                            'gender', 'imputers\imputers_train\gender_imputer_model')

2021-05-19 15:36:50,529 [INFO]  NumExpr defaulting to 6 threads.
2021-05-19 15:36:56,531 [INFO]  
========== start: fit model
2021-05-19 15:36:56,532 [WARNING]  Already bound, ignoring bind()
2021-05-19 15:36:58,817 [INFO]  Epoch[0] Batch [0-330]	Speed: 2605.51 samples/sec	cross-entropy=0.350056	gender-accuracy=0.904834
2021-05-19 15:37:00,848 [INFO]  Epoch[0] Train-cross-entropy=0.344722
2021-05-19 15:37:00,849 [INFO]  Epoch[0] Train-gender-accuracy=0.903357
2021-05-19 15:37:00,850 [INFO]  Epoch[0] Time cost=4.316
2021-05-19 15:37:00,947 [INFO]  Saved checkpoint to "imputers\imputers_train\gender_imputer_model\model-0000.params"
2021-05-19 15:37:01,217 [INFO]  Epoch[0] Validation-cross-entropy=0.315748
2021-05-19 15:37:01,218 [INFO]  Epoch[0] Validation-gender-accuracy=0.908784
2021-05-19 15:37:03,108 [INFO]  Epoch[1] Batch [0-330]	Speed: 2804.10 samples/sec	cross-entropy=0.320661	gender-accuracy=0.905400
2021-05-19 15:37:04,959 [INFO]  Epoch[1] Train-cross-entropy=0.325851
2021-05-19

In [39]:
imputed_gender_for_evaluation = gender_imputer_for_evaluation.predict(df_test)

In [40]:
imputed_gender_for_evaluation["gender"].value_counts()

Male      2624
Female     270
Other       35
Name: gender, dtype: int64

In [41]:
imputed_gender_for_evaluation["gender_imputed"].value_counts()

Male      2928
Female       1
Name: gender_imputed, dtype: int64

As it can be seen above, the prediction is biased due to the majority of male we have in our dataset. We did the evaluation just to show that this model will be not 100% precise. We won't repeat the evaluation for the other features in this notebook, so we will use as train dataset the whole dataset with no missing values.

In [42]:
gender_imputer = prediction(df_not_missing_gender, 
                            input_columns_for_gender, 
                            'gender', 'imputers\imputers_train\gender_imputer_model')

In [43]:
imputed_gender = gender_imputer.predict(dataFrame[dataFrame["gender"].isnull()])
imputed_gender['gender_imputed'].value_counts()

Male      4434
Female       4
Name: gender_imputed, dtype: int64

#### Enrolled University

In [44]:
df_not_missing_uni = dataFrame[dataFrame["enrolled_university"].isnull() == False]

input_columns_for_university = ['city', 'city_development_index', 'gender', 'relevent_experience',
       'education_level', 'major_discipline',
       'experience', 'company_size', 'company_type', 'last_new_job',
       'training_hours']

university_imputer = prediction(df_not_missing_uni, 
                            input_columns_for_university, 
                            'enrolled_university', 'imputers\\imputers_train\\university_imputer_model')

In [45]:
imputed_university = university_imputer.predict(dataFrame[dataFrame["enrolled_university"].isnull()])
imputed_university['enrolled_university_imputed'].value_counts()

no_enrollment       226
Full time course    106
Name: enrolled_university_imputed, dtype: int64

#### Education Level

In [46]:
df_not_missing_edu = dataFrame[dataFrame["education_level"].isnull() == False]

input_columns_for_education = ['city', 'city_development_index', 'gender', 'relevent_experience',
       'enrolled_university', 'major_discipline',
       'experience', 'company_size', 'company_type', 'last_new_job',
       'training_hours']

education_imputer = prediction(df_not_missing_edu, 
                            input_columns_for_education, 
                            'education_level', 'imputers\imputers_train\education_imputer_model')

In [47]:
imputed_education = education_imputer.predict(dataFrame[dataFrame["education_level"].isnull()])
imputed_education['education_level_imputed'].value_counts()

Graduate          362
Masters            29
Primary School      1
Name: education_level_imputed, dtype: int64

#### Major Discipline

In [48]:
df_not_missing_discipline = dataFrame[dataFrame["major_discipline"].isnull() == False]

input_columns_for_discipline = ['city', 'city_development_index', 'gender', 'relevent_experience',
       'enrolled_university', 'education_level',
       'experience', 'company_size', 'company_type', 'last_new_job',
       'training_hours']

discipline_imputer = prediction(df_not_missing_discipline, 
                            input_columns_for_discipline, 
                            'major_discipline', 'imputers\imputers_train\discipline_imputer_model')

In [49]:
imputed_discipline = discipline_imputer.predict(dataFrame[dataFrame["major_discipline"].isnull()])
imputed_discipline['major_discipline_imputed'].value_counts()

STEM        321
No Major     97
Other         1
Name: major_discipline_imputed, dtype: int64

In [50]:
imputed_discipline['major_discipline_imputed'] = np.where(imputed_discipline['major_discipline_imputed'] == "No Major", 
                                     'Other', 
                                     imputed_discipline['major_discipline_imputed'])

In [51]:
imputed_discipline['major_discipline_imputed'].value_counts()

STEM     321
Other     98
Name: major_discipline_imputed, dtype: int64

#### Experience

In [52]:
df_not_missing_experience = dataFrame[dataFrame["experience"].isnull() == False]

input_columns_for_experience = ['city', 'city_development_index', 'gender', 'relevent_experience',
       'enrolled_university', 'education_level',
       'major_discipline', 'company_size', 'company_type', 'last_new_job',
       'training_hours']

experience_imputer = prediction(df_not_missing_experience, 
                            input_columns_for_experience, 
                            'experience', 'imputers\imputers_train\experience_imputer_model')

In [53]:
imputed_experience = experience_imputer.predict(dataFrame[dataFrame['experience'].isnull()])
imputed_experience['experience_imputed'].value_counts()

>20    18
4      15
3       8
5       5
2       3
<1      1
1       1
6       1
7       1
Name: experience_imputed, dtype: int64

#### Company Size

In [54]:
df_not_missing_size = dataFrame[dataFrame["company_size"].isnull() == False]

input_columns_for_size = ['city', 'city_development_index', 'gender', 'relevent_experience',
       'enrolled_university', 'education_level',
       'major_discipline', 'experience', 'company_type', 'last_new_job',
       'training_hours']

size_imputer = prediction(df_not_missing_size, 
                            input_columns_for_size, 
                            'company_size', 'imputers\imputers_train\size_imputer_model')

In [55]:
imputed_size = size_imputer.predict(dataFrame[dataFrame['company_size'].isnull()])
imputed_size['company_size_imputed'].value_counts()

50-99        2622
100-500      1563
10000+        945
<10           522
10/49         140
1000-4999      68
500-999         2
5000-9999       2
Name: company_size_imputed, dtype: int64

#### Company Type

In [56]:
df_not_missing_type = dataFrame[dataFrame["company_type"].isnull() == False]

input_columns_for_type = ['city', 'city_development_index', 'gender', 'relevent_experience',
       'enrolled_university', 'education_level',
       'major_discipline', 'experience', 'company_size', 'last_new_job',
       'training_hours']

type_imputer = prediction(df_not_missing_type, 
                            input_columns_for_type, 
                            'company_type', 'imputers\\imputers_train\\type_imputer_model')

In [57]:
imputed_type = type_imputer.predict(dataFrame[dataFrame['company_type'].isnull()])
imputed_type['company_type_imputed'].value_counts()

Pvt Ltd                5722
Public Sector           322
Early Stage Startup      15
NGO                       6
Funded Startup            1
Name: company_type_imputed, dtype: int64

#### Last New Job

In [58]:
df_not_missing_lnj = dataFrame[dataFrame["last_new_job"].isnull() == False]

input_columns_for_lnj = ['city', 'city_development_index', 'gender', 'relevent_experience',
       'enrolled_university', 'education_level',
       'major_discipline', 'experience', 'company_size', 'company_type',
       'training_hours']

lnj_imputer = prediction(df_not_missing_lnj, 
                            input_columns_for_lnj, 
                            'last_new_job', 'imputers\imputers_train\lnj_imputer_model')

In [59]:
imputed_lnj = lnj_imputer.predict(dataFrame[dataFrame['last_new_job'].isnull()])
imputed_lnj['last_new_job_imputed'].value_counts()

1        255
never     49
>4        31
Name: last_new_job_imputed, dtype: int64

## Re-building 

In [60]:
def replaceMissingValues(original_df, imputed_df, feature_to_be_replaced, imputed_feature):
    
    '''
    It takes as input a dataset with missing values and the dataset obtained by
    imputation, the name of the feature to be filled (as string) and the name of the 
    imputed feature (as string) and gives as output the dataframe in which the missing 
    values of the feature in input are replaced with the imputed ones
    
    '''
    #merge original dataset with the imputed one
    original_df = original_df.merge(imputed_df[['enrollee_id', imputed_feature]],
                                    on='enrollee_id',how="left")
    
    #fill na values with imputed ones
    original_df[feature_to_be_replaced] = original_df[feature_to_be_replaced].fillna(original_df[imputed_feature])
    
    #drop the imputed feature
    original_df.drop(imputed_feature,inplace=True,axis=1)
    
    return original_df
    

In [61]:
new_df = dataFrame

In [62]:
new_df = replaceMissingValues(new_df, imputed_gender, 'gender', 'gender_imputed')
new_df = replaceMissingValues(new_df, imputed_university, 'enrolled_university', 'enrolled_university_imputed')
new_df = replaceMissingValues(new_df, imputed_education, 'education_level', 'education_level_imputed')
new_df = replaceMissingValues(new_df, imputed_discipline, 'major_discipline', 'major_discipline_imputed')
new_df = replaceMissingValues(new_df, imputed_experience, 'experience', 'experience_imputed')
new_df = replaceMissingValues(new_df, imputed_size, 'company_size', 'company_size_imputed')
new_df = replaceMissingValues(new_df, imputed_type, 'company_type', 'company_type_imputed')
new_df = replaceMissingValues(new_df, imputed_lnj, 'last_new_job', 'last_new_job_imputed')

In [70]:
missing_data = pd.DataFrame(new_df.isna().sum(), columns=[
                            'Number of missing values'])
missing_data['Percentage of missing values'] = missing_data['Number of missing values'].aggregate(
    lambda x: (x/dataFrame.shape[0])*100)
missing_data

Number of missing values  Percentage of missing values
enrollee_id                                    0                           0.0
city                                           0                           0.0
city_development_index                         0                           0.0
gender                                         0                           0.0
relevent_experience                            0                           0.0
enrolled_university                            0                           0.0
education_level                                0                           0.0
major_discipline                               0                           0.0
experience                                     0                           0.0
company_size                                   0                           0.0
company_type                                   0                           0.0
last_new_job                                   0                           0.0
training_hours                                 0                           0.0
target                                         0                           0.0

The cleaned dataset will be saved as a csv file named *cleaned_train.csv*.

In [72]:
new_df.to_csv(r'data\cleaned_train.csv', index = False)